velox/exec/Window.cpp

```diff
@@ -57,21 +57,71 @@ void Window::initialize() {
   windowNode_.reset();
 }
 
+namespace {
+void checkRowFrameBounds(const core::WindowNode::Frame& frame) {
+  auto frameBoundCheck = [&](const core::TypedExprPtr& frameValue) -> void {
+    if (frameValue == nullptr) {
+      return;
+    }
+
+    VELOX_USER_CHECK(
+        frameValue->type() == INTEGER() || frameValue->type() == BIGINT(),
+        "k frame bound must be INTEGER or BIGINT type");
+  };
+  frameBoundCheck(frame.startValue);
+  frameBoundCheck(frame.endValue);
+}
+
+void checkKRangeFrameBounds(
+    const std::shared_ptr<const core::WindowNode>& windowNode,
+    const core::WindowNode::Frame& frame,
+    const RowTypePtr& inputType) {
+  // For k Range frame bound:
+  // i) The order by needs to be a single column for bound comparisons.
+  // ii) The bounds values are pre-computed in the start(end)Value bound
+  // fields. So, start(end)Value bounds cannot be constants.
+  // iii) The frame bound column and the ORDER BY column must have
+  // the same type for correct comparisons.
+  VELOX_USER_CHECK_EQ(
+      windowNode->sortingKeys().size(),
+      1,
+      "Window frame of type RANGE PRECEDING or FOLLOWING requires single sorting key in ORDER BY.");
+
+  auto orderByType = windowNode->sortingKeys()[0]->type();
+  auto frameBoundCheck = [&](const core::TypedExprPtr& frameValue) -> void {
+    if (frameValue) {
+      auto frameChannel = exprToChannel(frameValue.get(), inputType);
+      VELOX_USER_CHECK_NE(
+          frameChannel,
+          kConstantChannel,
+          "Window frame of type RANGE does not support constant arguments");
+
+      auto frameType = inputType->childAt(frameChannel);
+      VELOX_USER_CHECK_EQ(
+          frameType,
+          orderByType,
+          "Window frame of type RANGE does not match types of the ORDER BY"
+          " and frame column");
+    }
+  };
+
+  frameBoundCheck(frame.startValue);
+  frameBoundCheck(frame.endValue);
+}
+
+}; // namespace
+
 Window::WindowFrame Window::createWindowFrame(
-    core::WindowNode::Frame frame,
+    const std::shared_ptr<const core::WindowNode>& windowNode,
+    const core::WindowNode::Frame& frame,
     const RowTypePtr& inputType) {
   if (frame.type == core::WindowNode::WindowType::kRows) {
-    auto frameBoundCheck = [&](const core::TypedExprPtr& frame) -> void {
-      if (frame == nullptr) {
-        return;
-      }
+    checkRowFrameBounds(frame);
+  }
 
-      VELOX_USER_CHECK(
-          frame->type() == INTEGER() || frame->type() == BIGINT(),
-          "k frame bound must be INTEGER or BIGINT type");
-    };
-    frameBoundCheck(frame.startValue);
-    frameBoundCheck(frame.endValue);
+  if (frame.type == core::WindowNode::WindowType::kRange &&
```

Comment:
Is it possible that both frame.startValue and frame.endValue are null? Can we move these checks inside checkKRangeFrameBounds?

Reply:
Yes, if frame is say RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW, then both startValue adn endValue are null. These have only startType and endType bound types. For a "k Range" frame atleast one of the bounds has an expression value. So I wrote it this way as the function validates "kRange" bounds only.